File: Packages/NotebarCore/Tests/NotebarCoreTests/PanelMachineTests.swift
import Testing
@testable import NotebarCore

@Suite("PanelMachine transitions")
struct PanelMachineTransitionTests {

    @Test("entering the trigger zone starts the dwell timer but does not expand")
    func enteringTriggerStartsDwell() {
        let (state, effects) = PanelMachine.reduce(.hidden, .cursorEnteredTrigger, .idle)
        #expect(state == .hidden)
        #expect(effects.contains(.startTimer(.edgeDwell)))
        #expect(effects.contains(.setPollRate(.active)))
    }

    @Test("leaving the trigger zone before dwell cancels the timer")
    func leavingTriggerCancelsDwell() {
        let (state, effects) = PanelMachine.reduce(.hidden, .cursorLeftTrigger, .idle)
        #expect(state == .hidden)
        #expect(effects.contains(.cancelTimer(.edgeDwell)))
        #expect(effects.contains(.setPollRate(.idle)))
    }

    @Test("dwell elapsing expands the panel")
    func dwellElapsedExpands() {
        let (state, effects) = PanelMachine.reduce(.hidden, .edgeDwellElapsed, .idle)
        #expect(state == .expanding)
        #expect(effects.contains(.showPanel))
    }

    @Test("animation finishing completes the expansion")
    func expandingCompletes() {
        let (state, _) = PanelMachine.reduce(.expanding, .animationFinished, .idle)
        #expect(state == .expanded)
    }

    @Test("leaving the panel starts the exit timer, it does not collapse immediately")
    func leavingPanelStartsExitTimer() {
        let (state, effects) = PanelMachine.reduce(.expanded, .cursorLeftPanel, .idle)
        #expect(state == .expanded)
        #expect(effects.contains(.startTimer(.exitDwell)))
        #expect(!effects.contains(.hidePanel))
    }

    @Test("returning to the panel cancels a pending collapse")
    func returningCancelsExitTimer() {
        let (state, effects) = PanelMachine.reduce(.expanded, .cursorEnteredPanel, .idle)
        #expect(state == .expanded)
        #expect(effects.contains(.cancelTimer(.exitDwell)))
    }

    @Test("exit dwell elapsing collapses the panel")
    func exitDwellCollapses() {
        let (state, effects) = PanelMachine.reduce(.expanded, .exitDwellElapsed, .idle)
        #expect(state == .collapsing)
        #expect(effects.contains(.hidePanel))
    }

    @Test("collapse animation finishing returns to hidden and slows polling")
    func collapseCompletes() {
        let (state, effects) = PanelMachine.reduce(.collapsing, .animationFinished, .idle)
        #expect(state == .hidden)
        #expect(effects.contains(.setPollRate(.idle)))
    }

    @Test("re-entering during collapse reverses it")
    func reEntryDuringCollapseReverses() {
        let (state, effects) = PanelMachine.reduce(.collapsing, .cursorEnteredPanel, .idle)
        #expect(state == .expanding)
        #expect(effects.contains(.showPanel))
    }

    @Test("escape collapses even when pinned")
    func escapeAlwaysCollapses() {
        let pinned = PanelContext(isPinned: true)
        let (state, effects) = PanelMachine.reduce(.expanded, .escapePressed, pinned)
        #expect(state == .collapsing)
        #expect(effects.contains(.hidePanel))
    }

    @Test("toggle opens when hidden and closes when expanded")
    func toggleFlips() {
        let (opened, openEffects) = PanelMachine.reduce(.hidden, .toggleRequested, .idle)
        #expect(opened == .expanding)
        #expect(openEffects.contains(.showPanel))

        let (closed, closeEffects) = PanelMachine.reduce(.expanded, .toggleRequested, .idle)
        #expect(closed == .collapsing)
        #expect(closeEffects.contains(.hidePanel))
    }

    @Test("unhandled pairs are inert")
    func unhandledPairsAreInert() {
        let (state, effects) = PanelMachine.reduce(.hidden, .exitDwellElapsed, .idle)
        #expect(state == .hidden)
        #expect(effects.isEmpty)
    }
}

File: windows/Notebar.Core.Tests/PanelMachineTests.cs
using Notebar.Core.Panel;
using Xunit;

namespace Notebar.Core.Tests;

public class PanelMachineTests
{
    private static (PanelState, IReadOnlyList<PanelEffect>) Reduce(
        PanelState state, PanelEvent evt, PanelContext? context = null) =>
        PanelMachine.Reduce(state, evt, context ?? PanelContext.Idle);

    [Fact]
    public void EnteringTriggerStartsDwell()
    {
        var (state, effects) = Reduce(PanelState.Hidden, PanelEvent.CursorEnteredTrigger);
        Assert.Equal(PanelState.Hidden, state);
        Assert.Equal(new PanelEffect[]
        {
            new PanelEffect.StartTimer(PanelTimerKind.EdgeDwell),
            new PanelEffect.SetPollRate(PollRate.Active),
        }, effects);
    }

    [Fact]
    public void LeavingTriggerCancelsDwell()
    {
        var (state, effects) = Reduce(PanelState.Hidden, PanelEvent.CursorLeftTrigger);
        Assert.Equal(PanelState.Hidden, state);
        Assert.Equal(new PanelEffect[]
        {
            new PanelEffect.CancelTimer(PanelTimerKind.EdgeDwell),
            new PanelEffect.SetPollRate(PollRate.Idle),
        }, effects);
    }

    [Fact]
    public void DwellElapsedExpands()
    {
        var (state, effects) = Reduce(PanelState.Hidden, PanelEvent.EdgeDwellElapsed);
        Assert.Equal(PanelState.Expanding, state);
        Assert.Equal(new PanelEffect[] { new PanelEffect.ShowPanel() }, effects);
    }

    [Fact]
    public void ExpandingCompletes()
    {
        var (state, effects) = Reduce(PanelState.Expanding, PanelEvent.AnimationFinished);
        Assert.Equal(PanelState.Expanded, state);
        Assert.Empty(effects);
    }

    /// Leaving an open panel only *arms* the collapse. Whether it is allowed to
    /// fire is decided when the timer elapses, against a fresh context.
    [Fact]
    public void LeavingPanelStartsExitTimer()
    {
        var (state, effects) = Reduce(PanelState.Expanded, PanelEvent.CursorLeftPanel);
        Assert.Equal(PanelState.Expanded, state);
        Assert.Equal(new PanelEffect[]
        {
            new PanelEffect.StartTimer(PanelTimerKind.ExitDwell),
        }, effects);
    }

    [Fact]
    public void ReturningCancelsExitTimer()
    {
        var (state, effects) = Reduce(PanelState.Expanded, PanelEvent.CursorEnteredPanel);
        Assert.Equal(PanelState.Expanded, state);
        Assert.Equal(new PanelEffect[]
        {
            new PanelEffect.CancelTimer(PanelTimerKind.ExitDwell),
        }, effects);
    }

    [Fact]
    public void ExitDwellCollapses()
    {
        var (state, effects) = Reduce(PanelState.Expanded, PanelEvent.ExitDwellElapsed);
        Assert.Equal(PanelState.Collapsing, state);
        Assert.Equal(new PanelEffect[] { new PanelEffect.HidePanel() }, effects);
    }

    [Fact]
    public void CollapseCompletes()
    {
        var (state, effects) = Reduce(PanelState.Collapsing, PanelEvent.AnimationFinished);
        Assert.Equal(PanelState.Hidden, state);
        Assert.Equal(new PanelEffect[] { new PanelEffect.SetPollRate(PollRate.Idle) }, effects);
    }

    /// Cursor came back mid-collapse: reverse without touching Hidden.
    [Fact]
    public void ReEntryDuringCollapseReverses()
    {
        foreach (var evt in new[] { PanelEvent.CursorEnteredPanel, PanelEvent.CursorEnteredTrigger })
        {
            var (state, effects) = Reduce(PanelState.Collapsing, evt);
            Assert.Equal(PanelState.Expanding, state);
            Assert.Equal(new PanelEffect[] { new PanelEffect.ShowPanel() }, effects);
        }
    }

    /// Escape overrides every suppression signal, pinning included.
    [Fact]
    public void EscapeAlwaysCollapses()
    {
        var pinned = PanelContext.Idle with { IsPinned = true, IsEditorFocused = true };
        foreach (var from in new[] { PanelState.Expanded, PanelState.Expanding })
        {
            var (state, effects) = Reduce(from, PanelEvent.EscapePressed, pinned);
            Assert.Equal(PanelState.Collapsing, state);
            Assert.Equal(new PanelEffect[]
            {
                new PanelEffect.HidePanel(),
                new PanelEffect.CancelTimer(PanelTimerKind.ExitDwell),
            }, effects);
        }
    }

    [Fact]
    public void ToggleFlips()
    {
        foreach (var from in new[] { PanelState.Hidden, PanelState.Collapsing })
        {
            var (state, effects) = Reduce(from, PanelEvent.ToggleRequested);
            Assert.Equal(PanelState.Expanding, state);
            Assert.Equal(new PanelEffect[]
            {
                new PanelEffect.ShowPanel(),
                new PanelEffect.SetPollRate(PollRate.Active),
            }, effects);
        }

        foreach (var from in new[] { PanelState.Expanded, PanelState.Expanding })
        {
            var (state, effects) = Reduce(from, PanelEvent.ToggleRequested);
            Assert.Equal(PanelState.Collapsing, state);
            Assert.Equal(new PanelEffect[] { new PanelEffect.HidePanel() }, effects);
        }
    }

    /// Anything not explicitly handled is inert: same state, no effects. This is
    /// what lets the controller fire events speculatively without checking state.
    [Fact]
    public void UnhandledPairsAreInert()
    {
        var (state, effects) = Reduce(PanelState.Hidden, PanelEvent.ExitDwellElapsed);
        Assert.Equal(PanelState.Hidden, state);
        Assert.Empty(effects);

        (state, effects) = Reduce(PanelState.Expanded, PanelEvent.EdgeDwellElapsed);
        Assert.Equal(PanelState.Expanded, state);
        Assert.Empty(effects);

        (state, effects) = Reduce(PanelState.Collapsing, PanelEvent.EscapePressed);
        Assert.Equal(PanelState.Collapsing, state);
        Assert.Empty(effects);
    }
}
